File: src/calculator/hifo.py
"""
HIFO (Highest In, First Out) lot selection.

Consumes the lots with the highest cost-per-unit first.
This minimizes taxable gain (or maximizes deductible loss) and is
generally the most tax-efficient method for appreciating assets.

Note: HIFO is a specific identification method recognized by the IRS for
crypto assets.  Taxpayers must be able to specifically identify the lots
(i.e., have records of cost basis per lot).
"""
from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from src.calculator.lots import Disposal, LotConsumption, LotManager, holding_period


def consume_hifo(
    lot_manager: LotManager,
    token: str,
    amount: Decimal,
    proceeds_usd: Decimal,
    disposal_date: datetime,
    tx_hash: str,
) -> Disposal:
    """
    Consume `amount` of `token` using HIFO order and return a Disposal.

    Lots with the highest cost-per-unit are consumed first.

    Args:
        lot_manager:   the shared in-memory lot book
        token:         e.g. "ETH"
        amount:        units being disposed
        proceeds_usd:  total USD received / fair-market-value at disposal
        disposal_date: datetime of the disposal event
        tx_hash:       source transaction

    Returns:
        Disposal with gain_loss_usd, holding_period, and lots_consumed populated.

    Raises:
        ValueError: if there are insufficient lots.
    """
    consumptions, total_cost = lot_manager.consume(
        token=token,
        amount=amount,
        disposal_date=disposal_date,
        order="hifo",
        method="HIFO",
    )

    gain_loss = proceeds_usd - total_cost
    hp = _blended_holding_period(consumptions, disposal_date, amount)

    return Disposal(
        date=disposal_date,
        token=token,
        amount=amount,
        proceeds_usd=proceeds_usd,
        cost_basis_usd=total_cost,
        gain_loss_usd=gain_loss,
        holding_period=hp,
        method="HIFO",
        lots_consumed=consumptions,
        tx_hash=tx_hash,
    )


def _blended_holding_period(
    consumptions: list[LotConsumption],
    disposal_date: datetime,
    total_amount: Decimal,
) -> str:
    """Return dominant holding period (long-term if >= 50% of amount is long-term)."""
    if not consumptions or total_amount == Decimal("0"):
        return "short-term"

    long_term_amount = Decimal("0")
    for c in consumptions:
        hp = holding_period(c.acquisition_date, disposal_date)
        if hp == "long-term":
            long_term_amount += c.amount_consumed

    if long_term_amount / total_amount >= Decimal("0.5"):
        return "long-term"
    return "short-term"

File: frontend/src/engine/calculator/hifo.ts
/**
 * HIFO (Highest In, First Out) lot consumption.
 *
 * Consumes lots with the highest cost-per-unit first (sorted by costPerUnit descending).
 * This minimizes taxable gain (or maximizes deductible loss) and is generally the
 * most tax-efficient method for appreciating assets.
 *
 * Ported from: src/calculator/hifo.py
 */

import Decimal from 'decimal.js';
import type { TaxLot, LotConsumption } from '../types';

/**
 * Consume `amount` of a token using HIFO order.
 *
 * Lots with the highest cost-per-unit are consumed first.
 * Lots are MUTATED in-place: `lot.remaining` is decreased.
 *
 * @param lots   Open lots for the token (remaining > 0). Will be sorted internally.
 * @param amount Units to consume.
 * @returns Array of LotConsumption records describing which lots were used.
 * @throws Error if insufficient lots exist.
 */
export function consumeHifo(
  lots: TaxLot[],
  amount: Decimal,
): LotConsumption[] {
  const EPSILON = new Decimal('0.00000001');

  // Sort by cost-per-unit descending (highest first).
  const sorted = [...lots].sort((a, b) => {
    const cpuA = a.amount.isZero()
      ? new Decimal('0')
      : a.costBasisUsd.dividedBy(a.amount);
    const cpuB = b.amount.isZero()
      ? new Decimal('0')
      : b.costBasisUsd.dividedBy(b.amount);
    // descending: b - a
    if (cpuB.greaterThan(cpuA)) return 1;
    if (cpuB.lessThan(cpuA)) return -1;
    return 0;
  });

  const consumptions: LotConsumption[] = [];
  let remainingToConsume = amount;

  for (const lot of sorted) {
    if (remainingToConsume.lessThanOrEqualTo(new Decimal('0'))) break;

    const costPerUnit = lot.amount.isZero()
      ? new Decimal('0')
      : lot.costBasisUsd.dividedBy(lot.amount);

    const toTake = Decimal.min(lot.remaining, remainingToConsume);
    const costTaken = toTake.times(costPerUnit);

    // Mutate lot in place
    lot.remaining = lot.remaining.minus(toTake);
    remainingToConsume = remainingToConsume.minus(toTake);

    consumptions.push({
      lotId: lot.id,
      token: lot.token,
      amountConsumed: toTake,
      costBasisConsumed: costTaken,
      acquisitionDate: lot.acquisitionDate,
    });
  }

  if (remainingToConsume.greaterThan(EPSILON)) {
    throw new Error(
      `Insufficient lots for ${lots[0]?.token ?? 'unknown'}: needed ${amount.toString()}, ` +
      `short by ${remainingToConsume.toString()}`
    );
  }

  return consumptions;
}
